apps/src/applab/applab.js

```diff
@@ -801,29 +807,181 @@ Applab.init = function(config) {
     if (codeModeButton) {
       dom.addClickTouchEvent(codeModeButton, Applab.onCodeModeButton);
     }
-    var designModeAddButton = document.getElementById('designModeAddButton');
-    if (designModeAddButton) {
-      dom.addClickTouchEvent(designModeAddButton, Applab.onDesignModeAddButton);
-    }
-    var designModeAddInput = document.getElementById('designModeAddInput');
-    if (designModeAddInput) {
-      dom.addClickTouchEvent(designModeAddInput, Applab.onDesignModeAddInput);
-    }
-    var designModeAddLabel = document.getElementById('designModeAddLabel');
-    if (designModeAddLabel) {
-      dom.addClickTouchEvent(designModeAddLabel, Applab.onDesignModeAddLabel);
-    }
     var designModeClear = document.getElementById('designModeClear');
     if (designModeClear) {
       dom.addClickTouchEvent(designModeClear, Applab.onDesignModeClear);
     }
 
+    // Allow elements to be dragged and dropped from the design mode
+    // element tray to the play space.
+    if (window.$) {
+      $('.new-design-element').draggable({
+        containment:"#codeApp",
+        helper:"clone",
+        appendTo:"#codeApp",
+        revert: 'invalid',
+        zIndex: 2,
+        start: function() {
+          studioApp.resetButtonClick();
+        }
+      });
+      var scale = vizAppWidth / Applab.appWidth;
+      var gridSize = 20;
+      $('#visualization').droppable({
+        accept: '.new-design-element',
+        drop: function (event, ui) {
+          var elementType = ui.draggable[0].dataset.elementType;
+
+          var left = ui.position.left / scale;
+          left = Math.round(left - left % gridSize);
+          var top = ui.position.top / scale;
+          top = Math.round(top - top % gridSize);
+
+          Applab.createElement(elementType, left, top);
+        }
+      });
+    }
+
   }
 
   user = {applabUserId: config.applabUserId};
 };
 
 /**
+ * The types of acceptable HTML elements in the levelHtml.
+ * @type {{BUTTON: string, LABEL: string, INPUT: string}}
+ */
+var ElementType = {
+  BUTTON: 'button',
+  LABEL: 'label',
+  INPUT: 'input'
+};
+Applab.ElementType = ElementType;
+
+/**
+ * Returns an element id with the given prefix which is unused within
+ * the levelHtml.
+ * @param {string} prefix
+ * @returns {string}
+ */
+Applab.getUnusedElementId = function (prefix) {
+  var divApplab = $('#divApplab');
+  for (var i = 1; i < 100; i++) {
```

Comment:
the general problem with the nextId pattern is that the the student's HTML could start with arbitrary Ids in it already. if button2, button3 and button9 exist already, then the only sensible value for nextId (or nextId['button'] would be 10, which seems like a pain to compute.

I chose a limit because eventually giving up seems better than entering an infinite loop. I chose 100 because I thought that would be high enough that no one would manually add more than that number of the same type of element using the Design Mode. In this case I need to return something and *prefix* seems as good a choice as any. The student can always edit the id if they actually plan to attach an event listener to their 100th element of the same type.

All of that said I realize this is a bit hacky and I'm open to specific suggestions. change limit to 1000? or just a while loop and trust it will end?